Tsx doc (#1)
hector-cao/libvirt: docs/formatcaps.rst

    </guest>

  </capabilities>

Host CPU model and features
~~~~~~~~~~~~~~~~~

> As desbcribed in the (`Host capabilities`_) section, libvirt exposes to users the list of Host CPU features. libvirt has a special way to expose this list, instead of providing the full set of features, libvirt specifies a CPU model name as baseline and additionnal features on top of it. 

Reviewer: ```suggestion
As desbcribed in the (`Host capabilities`_) section, libvirt exposes to users the list of Host CPU features. Libvirt has a special way to expose this list: Instead of providing the full - and thereby often very long - set of features, libvirt specifies a CPU model name as baseline and additional features on top of it. 
```

Reviewer: Typo's and text suggestions

Example:

::

  <capabilities>

    <host>
      <uuid>7b55704c-29f4-11b2-a85c-9dc6ff50623f</uuid>
      <cpu>
        <arch>x86_64</arch>
        <model>Skylake-Client-noTSX-IBRS</model>
        <vendor>Intel</vendor>
        ...
        <feature name='ds'/>
        <feature name='acpi'/>
        <feature name='ss'/>
        <feature name='ht'/>
        <feature name='tm'/>

> 

Reviewer: ```suggestion
...
```

Reviewer: suggests that there are more, that would even allow you to shorten the list by a few if you want

> The ideal case would be the returned baseline CPU model matching exactly the physical machine and no additional feature being needed, for example, if you are running on an ``Icelake`` Server CPU, the returned CPU model name is ``Icelake-Server``. However, this ideal situation does not happen most of the time for 2 main reasons:

Reviewer: ```suggestion
The ideal case would be that the baseline CPU model definition matches exactly the CPU present in the system and no additional feature is needed to express the capabilities of the CPU. For example, if you are running on a Server CPU you bought as ``Icelake`` type, the returned CPU model name could be ``Icelake-Server``. However, this ideal situation does rarely happen, mainly for two reasons:
```

Reviewer: I was once told values under 10 maybe and under 4 for sure should be words - not sure if that is true, but it reads better.

> - Since it is not practical to have a database listing all known CPU models, libvirt only has a small list of baseline CPU model names (``usr/share/libvirt/cpu_map/``). It chooses the one that shares the greatest number of features (CPUID bits and MSR features) with the actual host physical machine CPU and then lists the remaining named features.

Reviewer: ```suggestion
- Manufacturers do not ship one (=1) type of CPU under a given name, there are various different SKUs with different features under the same name. Yet it is not practical to have a database listing all of those variants. Instead Libvirt only has a list of a baseline CPU model names - roughly one per generation - in ``/usr/share/libvirt/cpu_map/``. It chooses the named type that shares the greatest number of features (CPUID bits and MSR features) with the actual CPU present in the machine and then lists the remaining named features as difference to that known name.
```

> - Some features might be unavailable for various reasons (BIOS and kernel configuration). As a consequence, the list of detected features only matches an older CPU generation with some additional features (e.g., Intel ``Broadwell`` is detected on an Intel ``Icelake`` server). One typical example where this situation happens is related to the TSX mitigation[1]. As a mitigation to the TAA side channel attack, the linux kernel disables by default TSX and its 2 features, ``rtm`` and ``hle``. Since most popular Linux distros (Ubuntu, ...) keep this default behavior and these 2 features are disabled, libvirt is unable to display the correct CPU model name for CPU generations that are affected by this mitigation (Intel ``Icelake``, ``Sapphire Rapids``, ``Granite Rapids``, ..).

Reviewer: ```suggestion
- Some features might be in the hardware, but unavailable for various reasons (BIOS and kernel configuration, disabled for security, ...). One typical example where this situation happens is related to the TSX mitigation [1]. As a mitigation to the TAA side channel attack, the Linux kernel disables by default TSX and its 2 features, ``rtm`` and ``hle``. Since many Linux distributions keep this safer default behavior these 2 features appear as disabled.
```

> Even though some effort has been done to address these situations (``-noTSX`` variants are added to cover the missing TSX features) and offer users the ability to have the exact CPU model name they are running on, this effort is not complete, and users *should not* expect to have the matching CPU model name but should consider the returned CPU model name as a baseline to build the complete available feature set of the Host CPU.

Reviewer: ```suggestion

As a consequence, the list of detected features rarely match a named and defined CPU type.
Sometimes that just means that you'll get the right name, but still a long list of features enabled or disabled on top of it.
But in other times it might even lead to the situation that a different named CPU type, usually older CPU generation, is closer to the features libvirt finds in the CPU present in the system.
In that cases it is closer to express the capabilities via an older name e.g. ``Broadwell`` plus some features than calling it ``Icelake`` with many more features disabled.
Due to that Libvirt is unable to display the correct CPU model name for some CPU generations, but that is fine - the purpose is not to confirm what generation branding the chip was sold by, but instead the shortest name +/- features to express its capabilities.

Some effort has been done to address these situations (like ``-noTSX`` variants are added to cover the missing TSX features mentioned above) and offer users the ability to more often see the CPU model name they expect, but this can never be fully complete. Therefore users *should not* expect to have the reported CPU model name to have any implications other than that of a named baseline to build the complete available feature set of the Host CPU.
```

    </guest>

  </capabilities>

Host CPU model and features
~~~~~~~~~~~~~~~~~

> As desbcribed in the (`Host capabilities`_) section, libvirt exposes to users the list of Host CPU features. Libvirt has a special way to expose this list: Instead of providing the full - and thereby often very long - set of features, libvirt specifies a CPU model name as baseline and additional features on top of it. 

Author: ```suggestion
As described in the (`Host capabilities`_) section, libvirt exposes to users the list of Host CPU features. Libvirt has a special way to expose this list: Instead of providing the full - and thereby often very long - set of features, libvirt specifies a CPU model name as baseline and additional features on top of it. 
```

Example:

::

  <capabilities>

    <host>
      <uuid>7b55704c-29f4-11b2-a85c-9dc6ff50623f</uuid>
      <cpu>
        <arch>x86_64</arch>
        <model>Skylake-Client-noTSX-IBRS</model>
        <vendor>Intel</vendor>
        ...
        <feature name='ds'/>
        <feature name='acpi'/>
        <feature name='ss'/>
        <feature name='ht'/>
        <feature name='tm'/>

> ...

Reviewer: ```suggestion
        ...
```

to make sure it is in the code block

The ideal case would be that the baseline CPU model definition matches exactly the CPU present in the system and no additional feature is needed to express the capabilities of the CPU. For example, if you are running on a Server CPU you bought as ``Icelake`` type, the returned CPU model name could be ``Icelake-Server``. However, this ideal situation does rarely happen, mainly for two reasons:

> - Manufacturers do not ship one (=1) type of CPU under a given name, there are various different SKUs with different features under the same name. Yet it is not practical to have a database listing all of those variants. Instead Libvirt only has a list of a baseline CPU model names - roughly one per generation - in ``/usr/share/libvirt/cpu_map/``. It chooses the named type that shares the greatest number of features (CPUID bits and MSR features) with the actual CPU present in the machine and then lists the remaining named features as difference to that known name.

Reviewer: ```suggestion
- Manufacturers do not ship one (=1) type of CPU under a given name, there are various different SKUs with different features under the same name. Yet it is not practical to have a database listing all of those variants. Instead Libvirt only has a list of a baseline CPU model names - roughly one per generation - in ``/usr/share/libvirt/cpu_map/``.
```

moving this one down

- Some features might be in the hardware, but unavailable for various reasons (BIOS and kernel configuration, disabled for security, ...). One typical example where this situation happens is related to the TSX mitigation [1]. As a mitigation to the TAA side channel attack, the Linux kernel disables by default TSX and its 2 features, ``rtm`` and ``hle``. Since many Linux distributions keep this safer default behavior these 2 features appear as disabled.

> As a consequence, the list of detected features rarely match a named and defined CPU type.

Reviewer: ```suggestion
It chooses the named baseline model that shares the greatest number of features (CPUID bits and MSR features) with the actual CPU present in the machine and then lists the remaining named features as difference to that known name.
As a consequence, the list of detected features rarely is a perfect match to a baseline model name.
```

Sometimes that just means that you'll get the right name, but still a long list of features enabled or disabled on top of it.

> But in other times it might even lead to the situation that a different named CPU type, usually older CPU generation, is closer to the features libvirt finds in the CPU present in the system.

Reviewer: ```suggestion
But in other times it might even lead to the situation that a different named baseline model, usually an older CPU generation, is closer to the features libvirt finds in the CPU present in the system.
```

> In that cases it is closer to express the capabilities via an older name e.g. ``Broadwell`` plus some features than calling it ``Icelake`` with many more features disabled.

Author: @cpaelzer I don't think we can express disabled/missing features, we can only express "additional" features that is why the baseline is always older CPU gen

Reviewer: I expected it to be able to use <feature policy='disable' name='lahf_lm'/>, but if you are right then yes the sentence needs to change.

any yeah, If you can confirm please change the text
maybe like:

Since capabilities are expressed by baseline CPU model name plus the further features that need to be added, such differences might lead to the situation that a different named baseline model, usually older CPU generation, is selected to express the capabilities of CPU present in the system. For example it might use a ``Broadwell`` plus some features rather than calling it ``Icelake``.

> Due to that Libvirt is unable to display the correct CPU model name for some CPU generations, but that is fine - the purpose is not to confirm what generation branding the chip was sold by, but instead the shortest name +/- features to express its capabilities.

Reviewer: ```suggestion
Due to all that Libvirt might sometimes display the an unexpected CPU model name, but that is fine - the purpose is not to confirm what generation-branding the chip was sold by, but instead the shortest set of named baseline model +/- features to express its capabilities.
```

    </guest>

  </capabilities>

Host CPU model and features
~~~~~~~~~~~~~~~~~

As described in the (`Host capabilities`_) section, libvirt exposes to users the list of Host CPU features. Libvirt has a special way to expose this list: Instead of providing the full - and thereby often very long - set of features, libvirt specifies a CPU model name as baseline and additional features on top of it. 

Example:

::

  <capabilities>

    <host>
      <uuid>7b55704c-29f4-11b2-a85c-9dc6ff50623f</uuid>
      <cpu>
        <arch>x86_64</arch>
        <model>Skylake-Client-noTSX-IBRS</model>
        <vendor>Intel</vendor>
        ...
        <feature name='ds'/>
        <feature name='acpi'/>
        <feature name='ss'/>
        <feature name='ht'/>
        <feature name='tm'/>
        ...

> The ideal case would be that the baseline CPU model definition matches exactly the CPU present in the system and no additional feature is needed to express the capabilities of the CPU. For example, if you are running on a Server CPU you bought as ``Icelake`` type, the returned CPU model name could be ``Icelake-Server``. However, this ideal situation does rarely happen, mainly for two reasons:

Reviewer 2: ```suggestion
The ideal case would be that the baseline CPU model definition matches exactly the CPU present in the system and no additional feature is needed to express the capabilities of the CPU. For example, if you are running on a Server CPU you bought as ``Icelake`` type, the returned CPU model name could be ``Icelake-Server``. However, this ideal situation rarely happens, for two main reasons:
```

- Manufacturers do not ship one (=1) type of CPU under a given name, there are various different SKUs with different features under the same name. Yet it is not practical to have a database listing all of those variants. Instead Libvirt only has a list of a baseline CPU model names - roughly one per generation - in ``/usr/share/libvirt/cpu_map/``.

- Some features might be in the hardware, but unavailable for various reasons (BIOS and kernel configuration, disabled for security, ...). One typical example where this situation happens is related to the TSX mitigation [1]. As a mitigation to the TAA side channel attack, the Linux kernel disables by default TSX and its 2 features, ``rtm`` and ``hle``. Since many Linux distributions keep this safer default behavior these 2 features appear as disabled.

> It chooses the named baseline model that shares the greatest number of features (CPUID bits and MSR features) with the actual CPU present in the machine and then lists the remaining named features as difference to that known name.

Reviewer 2: ```suggestion
It chooses the named baseline model that shares the greatest number of features (CPUID bits and MSR features) with the actual CPU present in the machine and then lists the remaining named features as differences to that known name.
```

> As a consequence, the list of detected features rarely is a perfect match to a baseline model name.

Reviewer 2: ```suggestion
As a consequence, the list of detected features is rarely a perfect match to a baseline model name.
```

Sometimes that just means that you'll get the right name, but still a long list of features enabled or disabled on top of it.

> But in other times it might even lead to the situation that a different named baseline model, usually an older CPU generation, is closer to the features libvirt finds in the CPU present in the system.

Reviewer 2: ```suggestion
At other times it might even lead to a different named baseline model, usually an older CPU generation, being closer to the features libvirt finds in the CPU present in the system.
```
